### PR TITLE
Fix compatibility for 4.0.0-a1-dev

### PR DESCRIPTION
**Pull request by 3D-I**
#114

immagine

### CODE REVIEW

File: includes/functions_install.php [hunk 1]
```diff
@@ -136,10 +136,20 @@ function load_schema_31($install_path = '', $install_dbms = false)
 			include($phpbb_root_path . 'includes/constants.' . $phpEx);
 		}
 
-		$finder = new \phpbb\finder(new \phpbb\filesystem(), $phpbb_root_path, null, $phpEx);
+		// Check if equal or greater than 4.0@dev
+		if (file_exists($phpbb_root_path . 'vendor-ext'))
```
Comment (VSEphpbb) [inline includes/functions_install.php]: Instead of this, define a PHPBB_40 constant in the index, like the way we defined PHPBB_31 and PHPBB_32. Then check here if PHPBB_40 is defined

Comment (DavidIQ) [reply includes/functions_install.php]: I kind of want to avoid that as it's kind of a bad practice to be honest.

Comment (VSEphpbb) [reply includes/functions_install.php]: But that's literally how all of QuickInstall is managing all the different coding requirements between the multiple phpBB versions. We can't have exotic tests everywhere, instead we use a defined constant that can be used everywhere checks are needed. 

Also it's for consistency and continuity as well. If it's a bad practice, it's too late as it's already everywhere in here (and most of QI code is bad practice too IMO, like using phpBB 3.0's old template engine and syntax, but that's another issue).

Comment (DavidIQ) [reply includes/functions_install.php]: I'm aware of how the rest of the app is written and that there's this nonsense all over it. This is fine for now I guess.

Comment (VSEphpbb) [reply includes/functions_install.php]: I agree with you fully on the nonsense 😃

Comment (3D-I) [reply includes/functions_install.php]: The use of constants has been discussed within this PR
https://github.com/phpbb/quickinstall/pull/112
where I started following the "logic" of the actual code, for consistency.

At some point this app should be rewritten IMO, so I agree with David.

Comment (VSEphpbb) [reply includes/functions_install.php]: Yes, and it should continue to be followed here. It's the only way to find where we're making special accomodations for specific phpBB versions... by searching for where these constants are being used.

You can not simply start creating new tests based on arbitrary conditions to accomodate phpBB specific code. That makes it VERY difficult to find phpBB version specific code in this application.

Comment (3D-I) [reply includes/functions_install.php]: Of course, for better or for worse, we're all right, it seems, and I agree. I would like to know exactly what approach to follow then. Obviously I consider the use of all these constants a little illogical, we can argue that even the present method is otherwise illogical.

It is up to you, then, to choose.

Comment (VSEphpbb) [reply includes/functions_install.php]: constants

Comment (3D-I) [reply includes/functions_install.php]: Okay, I will push the changes so we can see if the new approach satisfies we all.

Comment (VSEphpbb) [reply includes/functions_install.php]: Just to expand on that: If anybody wants to come up with an alternative to these constants that actually works well, that's great but that's also a separate project for another PR. For this PR, and for the time being, we should continue handling these different version-specific code conditions the way the app currently does it, which is by these constants.

**Comment by 3D-I** [comment]
Just pushed the use of constants.. NOT tested.

File: index.php [hunk 2]
```diff
@@ -153,6 +153,18 @@
 	define('PHPBB_32', true);
 }
 
+// Check if greater than 3.2
+if (!file_exists($phpbb_root_path . 'phpbb\db\driver\sqlite.' . $phpEx))
```
Comment (VSEphpbb) [inline index.php]: Can you please revert this change. I was OK with this as it was more about choosing the right sqlite class, and is not actually about phpBB versions (also 3.2 does not have the old sqlite class either so it's a false test for greater than 3.2).

This PR should just be about the 4.0 change please.

Comment (3D-I) [reply index.php]: Not a prob. 
Just a note, '!file_exists' means if the file (old sqlite class) does NOT exist.
Therefore the test is correct IMO.

Comment (VSEphpbb) [reply index.php]: The old sqlite class/file does not exist in 3.2.
https://github.com/phpbb/phpbb/tree/3.2.x/phpBB/phpbb/db/driver

Comment (3D-I) [reply index.php]: I know..

`if (!file_exists($phpbb_root_path . 'phpbb\db\driver\sqlite.' . $phpEx))`

IF the file does NOT exists then it is 3.3 or greater.

Comment (3D-I) [reply index.php]: Notice the `!`

Comment (VSEphpbb) [reply index.php]: Yes. And it does NOT exist in 3.2.x just like it does NOT exist in 3.3.x. It only exists in 3.1.

Comment (DavidIQ) [reply index.php]: Probably should have done a `class_exists` check like I had initially suggested in the originating pull request instead of a `file_exists` check. As @VSEphpbb mentioned though, this change should be reverted from this pr.

Comment (3D-I) [reply index.php]: What's the difference between the master branch and this one?

https://github.com/phpbb/quickinstall/blob/37fa565e6872d60ed06d47c12b8340b37f1a929c/includes/functions_install.php#L143-L151

Comment (3D-I) [reply index.php]: Oky I will revert it.

Comment (3D-I) [reply index.php]: @VSEphpbb Now I am seeing what you mean :+1:

File: includes/functions_install.php [hunk 3]
```diff
@@ -136,10 +136,20 @@ function load_schema_31($install_path = '', $install_dbms = false)
...
+		}
+		else
+		{
+			$finder = new \phpbb\finder(new \phpbb\filesystem(), $phpbb_root_path, null, $phpEx);
+		}
+
 		$classes = $finder->core_path('phpbb/db/migration/data/')
 			->get_classes();
 
+		// Check if greater than 3.2
```
Comment (VSEphpbb) [inline includes/functions_install.php]: Please remove this comment (and you can also remove the one above, it's not necessary as the test for PHPBB_40 is self-explanatory)

File: index.php [hunk 4]
```diff
@@ -153,6 +153,12 @@
 	define('PHPBB_32', true);
 }
 
+// Check if equal or greater than 4.0@dev
```
Comment (VSEphpbb) [inline index.php]: Comment here not really needed (we're not using them above so why start now....or otherwise add comments above)
